Riot/Modules/Settings/DeactivateAccount/DeactivateAccountViewController.m

```diff
@@ -331,19 +304,78 @@ - (IBAction)forgetMessagesButtonAction:(UIButton*)sender
 
 - (IBAction)deactivateAccountButtonAction:(id)sender
 {
-    __weak typeof(self) weakSelf = self;
+    [self startLoading];
     
-    [self presentPasswordRequiredAlertWithSubmitHandler:^(NSString *password) {
-        
-        typeof(weakSelf) strongSelf = weakSelf;
+    MXWeakify(self);
+    [self.deactivateAccountService checkAuthenticationWithSuccess:^(enum DeactivateAccountAuthentication authentication, NSURL * _Nullable fallbackURL) {
+        MXStrongifyAndReturnIfNil(self);
         
-        if (strongSelf)
-        {
-            NSString *userId = strongSelf.mainSession.myUser.userId;
-            [strongSelf deactivateAccountWithUserId:userId andPassword:password eraseAllMessages:strongSelf.forgetMessageButton.isEnabled];
+        switch (authentication) {
+            case DeactivateAccountAuthenticationAuthenticated:
+                MXLogDebug(@"[DeactivateAccountViewController] Deactivation endpoint has already been authenticated. Continuing deactivation.")
+                [self.deactivateAccountService deactivateWithEraseAccount:self.forgetMessageButton.isEnabled];
+                break;
+            case DeactivateAccountAuthenticationRequiresPassword:
+                [self presentPasswordPrompt];
+                break;
+            case DeactivateAccountAuthenticationRequiresFallback:
+                if (fallbackURL) [self presentFallbackForURL:fallbackURL];
+                break;
         }
-        
-    } cancelHandler:nil];
+    } failure:^(NSError * _Nonnull error) {
+        MXStrongifyAndReturnIfNil(self);
+        [self handleError:error];
+    }];
+}
+
+#pragma mark - Password
+
+- (void)presentPasswordPrompt
+{
+    MXLogDebug(@"[DeactivateAccountViewController] Show password prompt.")
+    
+    MXWeakify(self);
+    [self presentPasswordRequiredAlertWithSubmitHandler:^(NSString *password) {
+        MXStrongifyAndReturnIfNil(self);
+        [self.deactivateAccountService deactivateWith:password eraseAccount:self.forgetMessageButton.isEnabled];
+    } cancelHandler:^() {
+        MXStrongifyAndReturnIfNil(self);
+        [self stopLoading];
+    }];
+}
+
+#pragma mark - Fallback
+
+- (void)presentFallbackForURL:(NSURL *)url
+{
+    MXLogDebug(@"[DeactivateAccountViewController] Show fallback for url: %@", url)
+    SFSafariViewController *safariViewController = [[SFSafariViewController alloc] initWithURL:url];
+    safariViewController.modalPresentationStyle = UIModalPresentationFormSheet;
+    safariViewController.delegate = self;
+    
+    [self presentViewController:safariViewController animated:true completion:nil];
```

Comment:
```suggestion
    [self presentViewController:safariViewController animated:YES completion:nil];
```

Reply:
Guess who hasn't written much objc in a while 🙃